src/token-providing-subscription.ts

```diff
@@ -1,147 +1,161 @@
-                }               
-            }
+export let createTokenProvidingStrategy: (
+  nextSubscribeStrategy: SubscribeStrategy,
+  logger: Logger,
+  tokenProvider?: TokenProvider
+) => SubscribeStrategy = (nextSubscribeStrategy, logger, tokenProvider) => {
+  // Token provider might not be provided. If missing, go straight to underlying subscribe strategy.
+  if (tokenProvider) {
+    return (listeners, headers) =>
+      new TokenProvidingSubscription(
+        logger,
+        listeners,
+        headers,
+        tokenProvider,
+        nextSubscribeStrategy
+      );
+  } else {
+    return (listeners, headers) =>
+      nextSubscribeStrategy(listeners, headers);
+  }
+}
 
-            class OpenSubscriptionState implements SubscriptionState {
-                constructor(private headers: ElementsHeaders, private underlyingSubscription: Subscription, private onTransition: (SubscriptionState) => void){
-                    logger.verbose(`TokenProvidingSubscription: transitioning to OpenSubscriptionState`);
-                    listeners.onOpen(headers);
-                }
+export interface TokenProvidingSubscriptionState {
+  subscribe(token: string, listeners: SubscriptionListeners): void;
+  unsubscribe(): void;
+}
 
-                unsubscribe(){
-                    this.underlyingSubscription.unsubscribe();
-                    this.onTransition(new EndedSubscriptionState());
-                }
+class TokenProvidingSubscription implements Subscription {
+  private state: TokenProvidingSubscriptionState;
+  private tokenPromise: Promise<any>;
+
+  constructor(
+    private logger: Logger,
+    private listeners: SubscribeStrategyListeners,
+    private headers: ElementsHeaders,
+    private tokenProvider: TokenProvider,
+    private nextSubscribeStrategy: SubscribeStrategy
+  ) {
+    this.state = new ActiveState(logger, headers, nextSubscribeStrategy, this.onTransition);
+    this.subscribe();
+  }
+
+  public unsubscribe = () => {
+    this.state.unsubscribe();
+    this.onTransition(new InactiveState(this.logger));
+  }
+
+  private subscribe(): void {
+    this.tokenPromise = this.tokenProvider.fetchToken()
+      .then(token => {
+        const existingListeners = Object.assign({}, this.listeners);
+        const listenersWithWrappedOnError = Object.assign(
+          this.listeners,
+          {
+            onEnd: (error: any) => {
+              this.onTransition(new InactiveState(this.logger));
+              existingListeners.onEnd(error);
+            },
+            onError: (error: any) => {
+              if (this.isTokenExpiredError(error)) {
+                this.tokenProvider.clearToken(token);
+                this.subscribe();
+              } else {
+                this.onTransition(new InactiveState(this.logger));
+                existingListeners.onError(error);
+              }
             }
+          }
+        );
+        this.state.subscribe(token, listenersWithWrappedOnError);
```

Comment:
I think constructing the listeners manually is more direct than using `Object.assign`:

```
this.state.subscribe(token, {
  onOpen: this.listeners.onOpen,
  onEvent: this.listeners.onEvent,
  onEnd: (error: any) => {
    this.onTransition(new InactiveState(this.logger));
    existingListeners.onEnd(error);
  },
  onError: (error: any) => {
    if (this.isTokenExpiredError(error)) {
      this.tokenProvider.clearToken(token);
      this.subscribe();
    } else {
      this.onTransition(new InactiveState(this.logger));
      existingListeners.onError(error);
    }
  }
});
```

I'm a bit surprised that it compiled without warnings. Maybe types are not strict enough.